rust/src/storage/mod.rs

```diff
@@ -109,13 +111,25 @@ pub fn parse_uri<'a>(path: &'a str) -> Result<Uri<'a>, UriError> {
                             return Err(UriError::MissingObjectBucket);
                         }
                     };
-                    let key = match path_parts.next() {
+                    let raw_key = match path_parts.next() {
                         Some(x) => x,
                         None => {
                             return Err(UriError::MissingObjectKey);
                         }
                     };
 
+                    let key = match raw_key.contains('/') {
+                        false => Cow::Borrowed(raw_key),
+                        true => {
+                            let key_cleaned = raw_key.split('/')
+                                                     .filter(|p| !p.is_empty())
+                                                     .collect::<Vec<& str>>()
+                                                     .join("/");
+                            Cow::Owned(key_cleaned)
+                        }
+                    };
```

Comment:
We can trust rust compiler to do the right optimization and simplify this block to:

```suggestion
                    let key =  raw_key.split('/')
                                      .filter(|p| !p.is_empty())
                                      .collect::<Vec<& str>>()
                                      .join("/");
```

Reply:
Right, it is simpler and more readable!